File: leetcode/15. 3Sum.ts
//https://leetcode.com/problems/3sum/
function threeSum(nums: number[]): number[][] {
    let map = new Map()

    let ordered = nums.sort((a,b) => a-b)

    for(let i = 0; i < ordered.length; i++) {

        let curr = getUniqueArr(ordered[i], ordered)
        map.set(ordered[i], curr)
    }

    console.log(map)

    let twoMap = new Map()

    console.log(ordered)

    for(let i = 0; i < ordered.length; i++) {

        let arr = map.get(nums[i])
        
    }


    return []

};

function getUniqueArr(num: number, nums: number[]): number[] {

    let res: number[] = []

    for(let i = 0; i < nums.length; i++) {
        if(nums[i] !== num) res.push(nums[i])
    }

    return res
}

function getTwoUniqueArr(num: number, map: Map<number, number[]>): any {

}

File: leetcode/15. 3sum.go
import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	res := [][]int{}

	for i := 0; i < len(nums)-2; i++ {

		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		left := i + 1
		right := len(nums) - 1

		for left < right {
			sum := nums[i] + nums[left] + nums[right]

			if sum == 0 {
				res = append(res, []int{nums[i], nums[left], nums[right]})
				right--

				for left < right && nums[right] == nums[right+1] {
					right--
				}
			} else if sum > 0 {
				right--
			} else {
				left++
			}
		}

	}

	return res
}
